### PR TITLE
Add negative ack redelivery backoff.

### PR DESCRIPTION
**Pull request by hanmz**
Master Issue: #12379 

### Motivation

Add negative ack redelivery backoff.

### Modifications

- add new `NegativeAckBackoff` interface
- expose `egativeAckRedeliveryBackoff` in ConsumerBulider
- add unit test case

### CODE REVIEW

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/NegativeAckRedeliveryBackoff.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+@InterfaceStability.Stable
+public interface NegativeAckRedeliveryBackoff extends Serializable  {
+    /**
+     * @param redeliveryCount indicates the number of times the message was redelivered
+     */
+    long next(int redeliveryCount);
+
+    long getMinNackTimeMs();
+
+    long getMaxNackTimeMs();
```
Comment (codelipenghui) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/NegativeAckRedeliveryBackoff.java:43]: I think we don't need these 2 methods for the `NegativeAckRedeliveryBackoff` interface, it should only for the implementations?

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/NegativeAckRedeliveryBackoff.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    interface Builder {
```
Comment (codelipenghui) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/NegativeAckRedeliveryBackoff.java:45]: Should provide in the implementations not the interface?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java [hunk 3]
```diff
@@ -514,10 +514,10 @@ public void reconsumeLaterCumulative(Message<?> message, long delayTime, TimeUni
...
     }
 
-    @Override
-    public void negativeAcknowledge(Message<?> message) {
-        negativeAcknowledge(message.getMessageId());
-    }
+//    @Override
+//    public void negativeAcknowledge(Message<?> message) {
+//        negativeAcknowledge(message.getMessageId());
+//    }
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java:520]: Typo here?

**Review by codelipenghui — CHANGES_REQUESTED** [review]

**Comment by wolfstudy** [comment]
ping @codelipenghui PTAL again, thanks

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java [hunk 4]
```diff
@@ -749,4 +749,10 @@
      * Default: null
      */
     ConsumerBuilder<T> messagePayloadProcessor(MessagePayloadProcessor payloadProcessor);
+
+    /**
+     * Notice: the negativeAckRedeliveryBackoff will not work with `consumer.negativeAcknowledge(MessageId messageId)`
+     * because we are not able to get the redelivery count from the message ID.
+     */
```
Comment (codelipenghui) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java:756]: It's better to provide an example for how to enable the redelivery backoff such as using the built-in exponential backoff

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/NegativeAcksTracker.java [hunk 5]
```diff
@@ -95,6 +105,40 @@ public synchronized void add(MessageId messageId) {
...
+    public synchronized void add(Message<?> message) {
+        if (negativeAckRedeliveryBackoff == null) {
+            add(message.getMessageId());
+            return;
+        }
+        add(message.getMessageId(), message.getRedeliveryCount());
+    }
+
+    private synchronized void add(MessageId messageId, int redeliveryCount) {
+        if (messageId instanceof TopicMessageIdImpl) {
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/NegativeAcksTracker.java:117]: The `TopicMessageIdImpl` will not reach here? The NegativeAcksTracker maintained in the internal consumers

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/NegativeAckRedeliveryExponentialBackoff.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    public static NegativeAckRedeliveryExponentialBackoff.NegativeAckRedeliveryExponentialBackoffBuilder builder() {
+        return new NegativeAckRedeliveryExponentialBackoff.NegativeAckRedeliveryExponentialBackoffBuilder();
+    }
+
+    @Override
+    public long next(int redeliveryCount) {
+        if (redeliveryCount <= 0) {
+            return minNackTimeMs;
+        }
+        return Math.min(Math.abs(minNackTimeMs << redeliveryCount), maxNackTimeMs);
```
Comment (Jason918) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/NegativeAckRedeliveryExponentialBackoff.java:46]: When redeliveryCount is large , `minNackTimeMs << redeliveryCount` could be smaller than `maxNackTimeMs`, or even be 0, which will be confusing.
For example, `next()`  returns 0 if minNackTimeMs=10000 and redeliveryCount=60.

**Review by codelipenghui — COMMENTED** [review]
The change looks good, could you please also help add an integration test?

**Review by Jason918 — CHANGES_REQUESTED** [review]

**Comment by Anonymitaet** [comment]
Confirmed with @wolfstudy, he will provide tech inputs for doc

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/NegativeAcksTracker.java (hunk 5 above)
Comment (hanmz) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/NegativeAcksTracker.java:117]: If this message from MultiTopicsConsumerImpl, messageId instance of TopicMessageIdImpl.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/NegativeAckRedeliveryExponentialBackoff.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    private final long maxNackTimeMs;
+    private final int maxBitShift;
+
+    private NegativeAckRedeliveryExponentialBackoff(long minNackTimeMs, long maxNackTimeMs) {
+        this.minNackTimeMs = minNackTimeMs;
+        this.maxNackTimeMs = maxNackTimeMs;
+
+        for (int i = 0; ; ) {
+            if (this.minNackTimeMs << ++i <= 0) {
+                System.out.println(i);
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/NegativeAckRedeliveryExponentialBackoff.java:39]: Remove this line.

**Review by codelipenghui — APPROVED** [review]
LGTM, just left a minor comment, PTAL.

**Comment by wolfstudy** [comment]
> Confirmed with @wolfstudy, he will provide tech inputs for doc

Ok, will add docs in next pull request.

**Review by wolfstudy — APPROVED** [review]
LGTM +1

**Review by Jason918 — APPROVED** [review]
LGTM

**Pull request merged**